### ISSUE TITLE
ClickOnce:  "This file is not signed." in MageUI.exe

### ISSUE THREAD

**Issue by dtivel** [dotnet/sign#562]
## Repro steps
1. Create a .NET 7.0 C# WinForms application project named WinFormsApp1.
1. Create a ClickOnce publishing profile without manifest signing.
1. Publish to `bin\publish`.
1. Compress the `publish` folder into ZIP file named `publish.clickonce`.
1. Create a file named `filelist.txt` with the following content:

   ```text
   **/WinFormsApp1.*
   **/setup.exe
   ```

1. Execute:

   ```text
   sign code azure-key-vault -kvu **** -kvc MyCertificate -kvm -d Description -u http://description.test -t http://timestamp.digicert.com -b E:\testpass -fl .\filelist.txt .\publish.clickonce
   ```

1. Rename `publish.clickonce` to `publish.zip` and extract all contents to a folder.
1. Launch `mageui.exe` (under `C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\mage.exe`).
1. In Mage UI, open `WinFormsApp1.application`

### Expected result

The manifest will show as validly signed.

### Actual result

The manifest is shown as unsigned.

![image](https://user-images.githubusercontent.com/12734758/212801139-da573eda-20bb-4419-9346-68d8e312a943.png)

### Observations

1. `mage.exe` reports files as being validly signed.  Open a command prompt and execute:

   ```text
   E:\testpass>"C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\mage.exe" -ver publish\publish\WinFormsApp1.application
   Manifest has a valid signature.

   E:\testpass>"C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.8 Tools\mage.exe" -ver "publish\publish\Application Files\WinFormsApp1_1_0_0_0\WinFormsApp1.dll.manifest"
   Manifest has a valid signature.
   ```

1. Launching `WinFormsApp1.application` reports the app as validly signed.

   ![image](https://user-images.githubusercontent.com/12734758/212800585-3c633201-6599-466f-bd18-72024bb49086.png)

1. The manifest file also displays as modified, and maybe it's the _modified_ version of the file which MageUI.exe is reporting as unsigned.
1. Installing the ClickOnce application and launching it succeeds without issue.

CC @clairernovotny

**Comment by dtivel** [comment]
The simple answer is that MageUI.exe does not support ClickOnce applications signed by Sign CLI, and that's to be expected.

There are 2 separate issues here:

1.  MageUI.exe always displays a newly opened file as modified.  This is because in some document initialization code, MageUI.exe sets a property:

    ```C#
    Document.UpdateInterval = 0
    ```

    ...and the setter for this property calls a `Changed` event, which always marks the document as modified.
    ```C#
    public int UpdateInterval
    {
        get { ... }
        set { ...; Changed(); }
    }

    private void Changed()
    {
        IsModified = true;
        // fire event
    }
    ```

    This feels to me like an issue with MageUI.exe.

2.  MageUI.exe cannot understand a ClickOnce app signed with SHA-256 as Sign CLI does.  MageUI.exe finds this XML element in a manifest file:
    ```XML
    <SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha256" />
    ```

    ...and passes the algorithm value to `CryptoConfig.CreateFromName(...)`.  However, there are 2 problems:

    a.  `http://www.w3.org/2000/09/xmldsig#rsa-sha256` is unknown to `CryptoConfig`.  `CryptoConfig` knows about `http://www.w3.org/2001/04/xmldsig-more#rsa-sha256` instead.  From https://www.w3.org/TR/xmldsig-core/#sec-Versions:

    > The http://www.w3.org/2000/09/xmldsig# (dsig:) namespace was introduced in the first edition of this specification.
    ...
    > This specification uses algorithm identifiers in the namespace http://www.w3.org/2001/04/xmldsig-more# that were originally coined in [[RFC6931](https://www.w3.org/TR/xmldsig-core/#bib-RFC6931)]

    Sign CLI uses the 2000/09 namespace because that's what Mage tools do:  [Sign CLI](https://github.com/dotnet/sign/blob/9d37b5131dd0a30bd56186086cd663a0d3875b4d/src/Sign.Core/Native/mansign2.cs#L76) and [dotnet-mage](https://github.com/dotnet/deployment-tools/blob/6f3b3469724731befcfd875d24a78a21bcb8b904/src/clickonce/MageCLI/Command.cs#L35).

    b.  You can't simply add the algorithm name mapping in machine.config because the type is not public.
    ```XML
    <mscorlib>
      <cryptographySettings>
        <cryptoNameMapping>
          <cryptoClasses>
            <cryptoClass XmlDsigSha256="System.Security.Cryptography.RSAPKCS1SHA256SignatureDescription, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" />
          </cryptoClasses>
          <nameEntry name="http://www.w3.org/2000/09/xmldsig#rsa-sha256"
                     class="XmlDsigSha256" />
        </cryptoNameMapping>
      </cryptographySettings>
    </mscorlib>
    ```

    In short, MageUI.exe would need to be updated to support the 2000/09 namespace.

**Comment by dtivel** [comment]
@clairernovotny, close as expected?
